tensorflow_addons/optimizers/discriminative_layer_training_test.py

```diff
@@ -0,0 +1,271 @@
+
+    def _test_equal_with_no_layer_lr(self, model_fn, loss, opt):
+        """confirm that discriminative learning is almost the same as regular learning"""
+        learning_rate = 0.01
+        model = model_fn()
+        model.compile(loss=loss, optimizer=opt(learning_rate))
+
+        model_lr = model_fn()
+        d_opt = DiscriminativeWrapper(
+            opt, model_lr, verbose=False, learning_rate=learning_rate
+        )
+        model_lr.compile(loss=loss, optimizer=d_opt)
+
+        self._assert_training_losses_are_close(model, model_lr)
+
+    def _test_equal_0_layer_lr_to_trainable_false(self, model_fn, loss, opt):
+        """confirm 0 lr_mult for the model is the same as model not trainable"""
+        learning_rate = 0.01
+        model = model_fn()
+        model.trainable = False
+        model.compile(loss=loss, optimizer=opt(learning_rate))
+
+        model_lr = model_fn()
+        model_lr.lr_mult = 0.0
+        d_opt = DiscriminativeWrapper(
+            opt, model_lr, verbose=False, learning_rate=learning_rate
+        )
+        model_lr.compile(loss=loss, optimizer=d_opt)
+
+        self._assert_training_losses_are_close(model, model_lr)
+
+    def _test_equal_half_layer_lr_to_half_lr_of_opt(self, model_fn, loss, opt):
+        """confirm 0.5 lr_mult for the model is the same as optim with 0.5 lr"""
+
+        mult = 0.5
+        learning_rate = 0.01
+        model = model_fn()
+        model.compile(loss=loss, optimizer=opt(learning_rate * mult))
+
+        model_lr = model_fn()
+        model_lr.lr_mult = mult
+        d_opt = DiscriminativeWrapper(
+            opt, model_lr, verbose=False, learning_rate=learning_rate
+        )
+        model_lr.compile(loss=loss, optimizer=d_opt)
+
+        self._assert_training_losses_are_close(model, model_lr)
+
+    def _test_loss_changes_over_time(self, model_fn, loss, opt):
+        """confirm that model trains with lower lr on specific layer"""
+
+        learning_rate = 0.01
+        model_lr = model_fn()
+        model_lr.layers[0].lr_mult = 0.01
+        d_opt = DiscriminativeWrapper(
+            opt, model_lr, verbose=False, learning_rate=learning_rate
+        )
+        model_lr.compile(loss=loss, optimizer=d_opt)
+
+        loss_values = get_losses(get_train_results(model_lr))
+        self.assertLess(loss_values[-1], loss_values[0])
+
+    def _run_tests_in_notebook(self):
+        for name, method in DiscriminativeLearningTest.__dict__.items():
+            if callable(method) and name[:4] == "test":
+                print("running test %s" % name)
+                method(self)
+
+
+def run_distributed(devices):
+    def decorator(f):
+        def decorated(self, *args, **kwargs):
+            logical_devices = devices
+            strategy = tf.distribute.MirroredStrategy(logical_devices)
+            with strategy.scope():
+                f(self, *args, **kwargs)
+
+        return decorated
+
+    return decorator
```

Comment:
Can we not use: https://github.com/tensorflow/addons/blob/master/tensorflow_addons/utils/test_utils.py#L112 ?

Please explain the details in why it wasn't sufficient and the following tests are setup this way. Perhaps we can update the utility if something is missing.

Reply:
I remember having issues with the distributed decorator in test_utils. I will revisit this and try to use the run distributed decorator

Reply:
This error was from an earlier commit when I tried to use the run distributed wrapper. I'll try it again later today. 

It appears that the CI runs fail when using the test utils run distributed because it generates virtual devices that cannot be modified after initialized. 

I may be using this function incorrectly. 

Traceback (most recent call last):
  File "/root/.cache/bazel/_bazel_root/e051f2f195071208ea7f081f88730f4f/execroot/__main__/bazel-out/k8-opt/bin/tensorflow_addons/optimizers/discriminative_layer_training_test.runfiles/__main__/tensorflow_addons/utils/test_utils.py", line 121, in decorated
    logical_devices = create_virtual_devices(num_devices)
  File "/root/.cache/bazel/_bazel_root/e051f2f195071208ea7f081f88730f4f/execroot/__main__/bazel-out/k8-opt/bin/tensorflow_addons/optimizers/discriminative_layer_training_test.runfiles/__main__/tensorflow_addons/utils/test_utils.py", line 88, in create_virtual_devices
    for _ in range(num_devices)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/framework/config.py", line 603, in set_logical_device_configuration
    context.context().set_logical_device_configuration(device, logical_devices)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/eager/context.py", line 1280, in set_logical_device_configuration
    "Virtual devices cannot be modified after being initialized")
RuntimeError: Virtual devices cannot be modified after being initialized

Reply:
error has now been resolved. it was caused by initializing a session before tf.test.main() ran.